### PR TITLE
always update scheduler observed generation when scheduling result being patched successfully

### LINKED ISSUE #4250: SchedulerObservedGeneration Not Updated When No Clusters Match LabelSelector
**What happened**:  

When I using `labelSelector` in `clusterAffinity` and remove labels on cluster so that no clusters match the selector, the scheduling result in binding would be patched as nil and the works would be removed, but the `schedulerObservedGeneration` would not be updated.  

**What you expected to happen**:  

After the scheduling result being removed, the `schedulerObservedGeneration` should also be updated.

**How to reproduce it (as minimally and precisely as possible)**:

1. Prepare a propagationpolicy and a deployment matches the policy:  
    ```yaml
    apiVersion: policy.karmada.io/v1alpha1
    kind: PropagationPolicy
    metadata:
      name: test-pp
      namespace: test
    spec:
      placement:
        clusterAffinity:
          labelSelector:
            matchLabels:
              cluster-labels: ok
      resourceSelectors:
      - apiVersion: apps/v1
        kind: Deployment
        name: test-deploy
    ```

1. Apply the policy and deployment, and label a cluster with `cluster-labels=ok`: 

    ```
    karmada:~/karmada/test$ kubectl apply -f test-pp.yaml
    propagationpolicy.policy.karmada.io/test-pp created

    karmada:~/karmada/test$ kubectl apply -f deploy.yaml
    deployment.apps/test-deployment created

    karmada:~/karmada/test$ kubectl label cluster cluster-1 cluster-labels=ok
    cluster.cluster.karmada.io/cluster-1 labeled

    karmada:~/karmada/test$ kubectl get rb -n test test-deployment-deployment -o jsonpath='{.metadata.uid}{"\n"}'
    d1b2571e-be6d-47af-8e83-80941154a579

    karmada:~/karmada/test$ kubectl get work -A -l "resourcebinding.karmada.io/uid=d1b2571e-be6d-47af-8e83-80941154a579"
    NAMESPACE              NAME                        APPLIED   AGE
    karmada-es-cluster-1   test-deployment-db6788b48   True      14m
    ```

    Then get the resourcebinding: 

    ```yaml
    apiVersion: work.karmada.io/v1alpha2
    kind: ResourceBinding
    metadata:
      annotations:
        policy.karmada.io/applied-placement: '{"clusterAffinity":{"labelSelector":{"matchLabels":{"cluster-labels":"ok"}}},"clusterTolerations":[{"key":"cluster.karmada.io/not-ready","operator":"Exists","effect":"NoExecute","tolerationSeconds":300},{"key":"cluster.karmada.io/unreachable","operator":"Exists","effect":"NoExecute","tolerationSeconds":300}]}'
        propagationpolicy.karmada.io/name: test-pp
        propagationpolicy.karmada.io/namespace: test
      creationTimestamp: "2023-11-16T06:15:16Z"
      finalizers:
      - karmada.io/binding-controller
      generation: 4
      labels:
        propagationpolicy.karmada.io/name: test-pp
        propagationpolicy.karmada.io/namespace: test
        propagationpolicy.karmada.io/uid: e16c5d5c-32b6-484a-990a-a8d48d61f7a2
      name: test-deployment-deployment
      namespace: test
      ownerReferences:
      - apiVersion: apps/v1
        blockOwnerDeletion: true
        controller: true
        kind: Deployment
        name: test-deployment
        uid: acfa5d48-d3f1-417b-bcb2-d965b12499f6
      resourceVersion: "910596169"
      selfLink: /apis/work.karmada.io/v1alpha2/namespaces/test/resourcebindings/test-deployment-deployment
      uid: d1b2571e-be6d-47af-8e83-80941154a579
    spec:
      clusters:
      - name: cluster-1
        replicas: 2
      placement:
        clusterAffinity:
          labelSelector:
            matchLabels:
              cluster-labels: ok
        clusterTolerations:
        - effect: NoExecute
          key: cluster.karmada.io/not-ready
          operator: Exists
          tolerationSeconds: 300
        - effect: NoExecute
          key: cluster.karmada.io/unreachable
          operator: Exists
          tolerationSeconds: 300
      replicaRequirements:
        resourceRequest:
          cpu: 100m
          memory: 1Gi
      replicas: 2
      resource:
        apiVersion: apps/v1
        kind: Deployment
        name: test-deployment
        namespace: test
        resourceVersion: "910595470"
        uid: acfa5d48-d3f1-417b-bcb2-d965b12499f6
      schedulerName: default-scheduler
    status:
      aggregatedStatus:
      - applied: true
        clusterName: cluster-1
        health: Unhealthy
        status:
          replicas: 2
          unavailableReplicas: 2
          updatedReplicas: 2
      conditions:
      - lastTransitionTime: "2023-11-16T06:15:51Z"
        message: Binding has been scheduled successfully.
        reason: Success
        status: "True"
        type: Scheduled
      - lastTransitionTime: "2023-11-16T06:15:51Z"
        message: All works have been successfully applied
        reason: FullyAppliedSuccess
        status: "True"
        type: FullyApplied
      schedulerObservedGeneration: 4
    ```

1. Then change the labels on cluster to `cluster-labels=ok`: 

    ```
    karmada:~/karmada/test$ kubectl label cluster cluster-1 cluster-labels=not_ok --overwrite
    cluster.cluster.karmada.io/cluster-1 labeled

    karmada:~/karmada/test$ kubectl get work -A -l "resourcebinding.karmada.io/uid=d1b2571e-be6d-47af-8e83-80941154a579"
    No resources found
    ```

    Then get the resourcebinding again:  

    ```yaml
    apiVersion: work.karmada.io/v1alpha2
    kind: ResourceBinding
    metadata:
      annotations:
        policy.karmada.io/applied-placement: '{"clusterAffinity":{"labelSelector":{"matchLabels":{"cluster-labels":"ok"}}},"clusterTolerations":[{"key":"cluster.karmada.io/not-ready","operator":"Exists","effect":"NoExecute","tolerationSeconds":300},{"key":"cluster.karmada.io/unreachable","operator":"Exists","effect":"NoExecute","tolerationSeconds":300}]}'
        propagationpolicy.karmada.io/name: test-pp
        propagationpolicy.karmada.io/namespace: test
      creationTimestamp: "2023-11-16T06:15:16Z"
      finalizers:
      - karmada.io/binding-controller
      generation: 5
      labels:
        propagationpolicy.karmada.io/name: test-pp
        propagationpolicy.karmada.io/namespace: test
        propagationpolicy.karmada.io/uid: e16c5d5c-32b6-484a-990a-a8d48d61f7a2
      name: test-deployment-deployment
      namespace: test
      ownerReferences:
      - apiVersion: apps/v1
        blockOwnerDeletion: true
        controller: true
        kind: Deployment
        name: test-deployment
        uid: acfa5d48-d3f1-417b-bcb2-d965b12499f6
      resourceVersion: "910604985"
      selfLink: /apis/work.karmada.io/v1alpha2/namespaces/test/resourcebindings/test-deployment-deployment
      uid: d1b2571e-be6d-47af-8e83-80941154a579
    spec:
      placement:
        clusterAffinity:
          labelSelector:
            matchLabels:
              cluster-labels: ok
        clusterTolerations:
        - effect: NoExecute
          key: cluster.karmada.io/not-ready
          operator: Exists
          tolerationSeconds: 300
        - effect: NoExecute
          key: cluster.karmada.io/unreachable
          operator: Exists
          tolerationSeconds: 300
      replicaRequirements:
        resourceRequest:
          cpu: 100m
          memory: 1Gi
      replicas: 2
      resource:
        apiVersion: apps/v1
        kind: Deployment
        name: test-deployment
        namespace: test
        resourceVersion: "910595470"
        uid: acfa5d48-d3f1-417b-bcb2-d965b12499f6
      schedulerName: default-scheduler
    status:
      conditions:
      - lastTransitionTime: "2023-11-16T06:25:10Z"
        message: '0/2 clusters are available: 1 cluster(s) did not match the placement
          cluster affinity constraint, 1 cluster(s) had untolerated taint {cluster.karmada.io/not-ready:NoSchedule}.'
        reason: NoClusterFit
        status: "False"
        type: Scheduled
      - lastTransitionTime: "2023-11-16T06:15:51Z"
        message: All works have been successfully applied
        reason: FullyAppliedSuccess
        status: "True"
        type: FullyApplied
      schedulerObservedGeneration: 4
    ```

**Anything else we need to know?**:

**Environment**:
- Karmada version:
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

### PR DESCRIPTION
**Pull request by lxtywypc**
**What type of PR is this?**

**What this PR does / why we need it**: 

**Which issue(s) this PR fixes**:
Fixes #4250 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmada-scheduler`: `schedulerObservedGeneration` would be always updated when scheduling result being patched successfully
```

### CODE REVIEW

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -745,14 +745,14 @@ func (s *Scheduler) establishEstimatorConnections() {
...
 	klog.V(4).Infof("Begin to patch status condition to ResourceBinding(%s/%s)", rb.Namespace, rb.Name)
 
 	updateRB := rb.DeepCopy()
 	meta.SetStatusCondition(&updateRB.Status.Conditions, newScheduledCondition)
-	// Postpone setting observed generation until schedule succeed, assume scheduler will retry and
-	// will succeed eventually.
-	if newScheduledCondition.Status == metav1.ConditionTrue {
+	// If an error during scheduling could be ignored, the scheduling result would have been patched successfully.
+	// The scheduler observed generation should be updated.
+	if ignoreErr {
```
Comment (XiShanYongYe-Chang) [inline pkg/scheduler/scheduler.go:755]: I feel that this judgment condition `newScheduledCondition.Status == metav1.ConditionTrue` also needs to be retained, the relationship between the two is OR. What do you think?

Comment (lxtywypc) [reply pkg/scheduler/scheduler.go:755]: https://github.com/ctripcloud/karmada/blob/48b8e9f9ba8a4e09a58524631aa596e84b8a8fb4/pkg/scheduler/helper.go#L95  

In fact, after `getConditionByError`, `ignoreErr` would be `true` when condition is `true`, so I only judge `ignoreErr` here.  

But it seems to be easily misunderstanding, I will fix it soon.

Comment (lxtywypc) [reply pkg/scheduler/scheduler.go]: Done.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -745,14 +745,14 @@ func (s *Scheduler) establishEstimatorConnections() {
...
-func patchBindingStatusCondition(karmadaClient karmadaclientset.Interface, rb *workv1alpha2.ResourceBinding, newScheduledCondition metav1.Condition) error {
+func patchBindingStatusCondition(karmadaClient karmadaclientset.Interface, rb *workv1alpha2.ResourceBinding, newScheduledCondition metav1.Condition, ignoreErr bool) error {
 	klog.V(4).Infof("Begin to patch status condition to ResourceBinding(%s/%s)", rb.Namespace, rb.Name)
 
 	updateRB := rb.DeepCopy()
 	meta.SetStatusCondition(&updateRB.Status.Conditions, newScheduledCondition)
-	// Postpone setting observed generation until schedule succeed, assume scheduler will retry and
-	// will succeed eventually.
-	if newScheduledCondition.Status == metav1.ConditionTrue {
+	// If an error during scheduling could be ignored, the scheduling result would have been patched successfully.
```
Comment (XiShanYongYe-Chang) [inline pkg/scheduler/scheduler.go:753]: New comments can be added under the existing comments.

Comment (lxtywypc) [reply pkg/scheduler/scheduler.go:755]: Done.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~
/lgtm
/cc @RainbowMango 

Do you think we need to cherry-pick this patch to the previous branch?

**Comment by lxtywypc** [comment]
> Do you think we need to cherry-pick this patch to the previous branch?  

This bug occurs after https://github.com/karmada-io/karmada/pull/3741, so we only need to cherry-pick it to v1.7.

**Review by RainbowMango — COMMENTED** [review]
/assign

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -424,7 +424,7 @@ func (s *Scheduler) doScheduleClusterBinding(name string) (err error) {
 func (s *Scheduler) scheduleResourceBinding(rb *workv1alpha2.ResourceBinding) (err error) {
 	defer func() {
 		condition, ignoreErr := getConditionByError(err)
-		if updateErr := patchBindingStatusCondition(s.KarmadaClient, rb, condition); updateErr != nil {
+		if updateErr := patchBindingStatusCondition(s.KarmadaClient, rb, condition, ignoreErr); updateErr != nil {
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:427]: Can anyone remind me of the root cause of this bug? 

cc @whitewindmills to take a look.

Comment (whitewindmills) [reply pkg/scheduler/scheduler.go:427]: In fact, the `NoClusterFit` reason does not represent a successful scheduling. When a new cluster meets the scheduling requirements, the binding in the `NoClusterFit` reason will be rescheduled.
In addition, let's understand the definition of the `schedulerObservedGeneration` field.
https://github.com/karmada-io/karmada/blob/186e137eecde2ca63140265c79fa6a5c7946ab5d/pkg/apis/work/v1alpha2/binding_types.go#L281-L285
In summary, I think the behavior mentioned in the issue(https://github.com/karmada-io/karmada/issues/4250) is expected.

Comment (lxtywypc) [reply pkg/scheduler/scheduler.go:427]: In my opinion, even if the `NoClusterFit` is not  a successful scheduling, the scheduling result quite has been updated and the workloads in member clusters would be removed, that means the `generation` has been 'observed' by scheduler, and the scheduler has finished its mission on this circumstance. So this field should be updated.

Comment (RainbowMango) [reply pkg/scheduler/scheduler.go:427]: > and the scheduler has finished its mission on this circumstance. So this field should be updated.

You are both right. The confusion is just because in the case of `NoClusterFit`, the scheduler would stop retrying.

The confusion comes from the unclear API definitions :( We can continue the discussion and try to clarify the API, hopefully we can make some updates to it.

Comment (XiShanYongYe-Chang) [reply pkg/scheduler/scheduler.go:427]: For Karmada-scheduler, there are two levels of meaning in the scheduling result. One meaning is whether the scheduling is completed, and the other meaning is whether the scheduling is successful. The current Condition only represents the second meaning, while the first meaning is handled in hidden code logic. The current PR fixes the issue where the hidden logic was not properly handled. 

I feel that it would be a good outcome to also reflect the first meaning at the API level so that the scheduler can have more input to determine whether to trigger scheduling again. 

However, the design consideration cycle of the API may be relatively long. Therefore, before that, the current problem will continue to exist. Therefore, my suggestion is to first promote the integration of the problem and then consider the issues related to Condition API design from a holistic perspective.
